### DISCUSSION TITLE
Are cookieStore instances unique per request?

### DISCUSSION THREAD

**Discussion by adamjmcgrath** [vercel/next.js#50321] [App Router]
If reading cookies was computationally expensive (say because I needed to decrypt the cookie) and I only wanted to do it once. Could I potentially cache that work per request against the `cookieStore` instance that is returned from invoking `cookies` from `next/headers` eg

```js
import { cookies } from 'next/headers';

const cache = new WeakMap();

const getSession() => {
  const cookieStore = cookies();
  let session = cache.get(cookieStore);
  if (!session) {
    session = expensiveDecryptCookie(cookieStore.get('session');
    cache.set(cookieStore, session);
  }
  return session;
}
```

Then you could call `getSession()` as frequently as you wanted during the lifecycle of the request and only decrypt the cookie once.

Is this possible? Can I safely rely on the cookieStore instance being unique per request and de-referenced at the end of the request lifecycle?

**Comment by jalvarado91** [comment]
Looks like [`cookies()` is part of a request store](https://github.com/vercel/next.js/blob/canary/packages/next/src/server/async-storage/request-async-storage-wrapper.ts#L95) that gets created per request instance in held on by the Next request cycle in `AsyncLocalStore` so it should be safe to assume the store reference will be unique per request, since AsyncLocalStorage should only be alive during the request lifetime. 

Curiously enough, I was digging through the implementation of the React `cache` function that's mentioned in the [Next docs under data fetching](https://github.com/vercel/next.js/blob/canary/packages/next/src/server/async-storage/request-async-storage-wrapper.ts#L95) because I was curious on what the heuristic for identity was used. Turns [to be basically a Map](https://github.com/facebook/react/blob/main/packages/react-server/src/ReactFlightServer.js#L366) that tracks the [function reference and arguments as keys](https://github.com/facebook/react/blob/e1ad4aa3615333009d76f947ff05ddeff01039c6/packages/react/src/ReactCache.js#L63-L66) internally. It's part of the react server request which is also resolved from AsyncLocalStorage, so for your example it might be enough to use the react cache to memoize your `getSession` function per request if you're within a RSC scope

**Comment by adamjmcgrath** [comment]
> so for your example it might be enough to use the react cache to memoize your getSession function per request if you're within a RSC scope

Yeah, I've done some testing locally and this seems to be the case. It's just [the example](https://nextjs.org/docs/app/building-your-application/data-fetching/caching#react-cache) `getUser` method feels like it is suggesting that the cache might persist across requests, otherwise why use a method that memoizes by user id for the example.
